### DISCUSSION TITLE
[RFC] Let's talk about bridge. v2.0.0

### DISCUSSION THREAD

**Discussion by solvedDev** [bridge-core/bridge.#244] [General]
# bridge. v2.0.0
bridge. v2.0.0 will hold a couple of big changes to the application. We believe these changes are necessary to future-proof the editor and make it appeal to an even broader group of users. 

## Project Structure:
We are looking to make substantial changes to how a project is structured. In order to keep compatibility with older projects, we will provide the option to automatically upgrade old projects to the new format.

### Terminology:
With bridge. v2.0.0, a project is not considered to be a single behavior pack anymore. A project consists of multiple packs, worlds and marketing material.

### Proposal:
```
📂 My Project
  📁 .git
  📁 bridge
  📁 marketing
  📂 packs
    📁 My BP
    📁 My RP
    📁 My SP
  📁 worlds
  🗄️ project.json
  🗄️ .gitignore
```

## Project Compiler:
Our goal is to load projects from a folder dedicated to bridge. projects. Upon changing the selected project, the compiler kicks in automatically, compiles the whole project and moves packs and worlds to the `com.mojang` folder so Minecraft can load them correctly. Upon making changes to your projects, bridge. will only update the files that changed. Packs and worlds will therefore automatically sync seamlessly between the two folders.

### Project Target Version:
We also want to allow users to lock the Minecraft version they want to develop for. Applying this lock will prevent developers from using features which aren't supported by the specified Minecraft version yet.

### Find & Replace:
Removing bridge.'s cache will allow us to build a global find & replace system for bridge.

### Motivation:
With a compiler based editor, we can get rid of bridge.'s cache system which enables better usability for teams. Additionally, we can build richer custom syntax features that allow developers and artists to be even more productive. 

## Tab System:
bridge.'s tab system will get a big overhaul to support exciting new development features. Just to name one example: We want to add live previews for BP entities, models and animations. This live preview will boot up in a separate tab and make use of the split-screen by default. Upon making changes to the code associated with a live preview, bridge. will refresh the preview to reflect the changes automatically.

## Testing Entities:
Jumping into Minecraft, booting up a world and testing certain behaviors can be pretty time consuming during development and if one of your changes breaks other behaviors, you may not even notice that until later in development which might cost you even more time debugging the issue. We want to give you the ability to program tests which ensure that things you code work and keep working. While this won't eliminate manual testing entirely, it will hopefully give you the ability to catch mistakes earlier. BP entity simulation (See: Tab System) can also make use of the backend code for testing entities.

## Plugin API 2.0:
Our new API is already partially available with bridge. v1.7.0 but we will continue development of it. 
bridge. v2.0.0 will include dynamic plugin loading. This means that a full reload of all plugins won't be required upon installation, activation or deactivation of plugins.

**Comment by Chikorita-Lover** [comment]
Dang, this is definitely something I really want to see happen. Very excited!

**Comment by ihategravel** [comment]
Coolio

**Reply by ihategravel2** [reply]
Bridge v2 is out now!
